ot/lp/dmmot.py

```diff
@@ -0,0 +1,344 @@
+
+    The algorithm accepts :math:`d` distributions (i.e., histograms)
+    :math:`a_{1}, \ldots, a_{d} \in \mathbb{R}_{+}^{n}` with :math:`e^{\prime}
+    a_{j}=1` for all :math:`j \in[d]`. Although the algorithm states that all
+    histograms have the same number of bins, the algorithm can be easily
+    adapted to accept as inputs :math:`a_{i} \in \mathbb{R}_{+}^{n_{i}}`
+    with :math:`n_{i} \neq n_{j}` [50].
+
+    The function solves the following optimization problem[51]:
+
+    .. math::
+        \begin{align}\begin{aligned}
+            \underset{\gamma\in\mathbb{R}^{n^{d}}_{+}} {\textrm{min}}
+            \sum_{i_1,\ldots,i_d} c(i_1,\ldots, i_d)\, \gamma(i_1,\ldots,i_d)
+            \quad \textrm{s.t.}
+            \sum_{i_2,\ldots,i_d} \gamma(i_1,\ldots,i_d) &= a_1(i_i),
+            (\forall i_1\in[n])\\
+            \qquad\vdots\\
+            \sum_{i_1,\ldots,i_{d-1}} \gamma(i_1,\ldots,i_d) &= a_{d}(i_{d}),
+            (\forall i_d\in[n]).
+            \end{aligned}
+        \end{align}
+
+
+    Parameters
+    ----------
+    A : nx.ndarray, shape (dim, n_hists)
+        The input ndarray containing distributions of n bins in d dimensions.
+    verbose : bool, optional
+        If True, print debugging information during execution. Default=False.
+    log : bool, optional
+        If True, record log. Default is False.
+
+    Returns
+    -------
+    obj : float
+        the value of the primal objective function evaluated at the solution.
+    log : dict
+        A dictionary containing the log of the discrete mmot problem:
+        - 'A': a dictionary that maps tuples of indices to the corresponding
+        primal variables. The tuples are the indices of the entries that are
+        set to their minimum value during the algorithm.
+        - 'primal objective': a float, the value of the objective function
+        evaluated at the solution.
+        - 'dual': a list of arrays, the dual variables corresponding to
+        the input arrays. The i-th element of the list is the dual variable
+        corresponding to the i-th dimension of the input arrays.
+        - 'dual objective': a float, the value of the dual objective function
+        evaluated at the solution.
+
+
+    References
+    ----------
+    .. [55] Ronak Mehta, Jeffery Kline, Vishnu Suresh Lokhande, Glenn Fung, &
+        Vikas Singh (2023). Efficient Discrete Multi Marginal Optimal
+        Transport Regularization. In The Eleventh International
+        Conference on Learning Representations.
+    .. [56] Jeffery Kline. Properties of the d-dimensional earth mover's
+        problem. Discrete Applied Mathematics, 265: 128-141, 2019.
+    .. [58] Leonid V Kantorovich. On the translocation of masses. Dokl. Akad.
+        Nauk SSSR, 37:227-229, 1942.
+
+    See Also
+    --------
+    ot.lp.dmmot_monge_1dgrid_optimize : Optimize the d-Dimensional Earth
+    Mover's Distance (d-MMOT)
+    """
+
+    nx = get_backend(A)
+    A_copy = nx.copy(A)
```

Comment:
```suggestion
    A_copy = A
```

We nee to keep teh original input to have a gradient